api/envoy/config/route/v3/route.proto

```diff
@@ -119,6 +120,18 @@ message RouteConfiguration {
   //   is not subject to data plane buffering controls.
   //
   google.protobuf.UInt32Value max_direct_response_body_size_bytes = 11;
+
+  // [#not-implemented-hide:]
+  // A list of plugins and their configurations which may be used by a
+  // :ref:`envoy_v3_api_field_config.route.v3.RouteAction.cluster_specifier_plugin`
+  // within the route. All *extension.name* fields in this list must be unique.
+  repeated ClusterSpecifierPlugin cluster_specifier_plugins = 12;
```

Comment:
shall we use `map<string, ClusterSpecifierPlugin>` here? That guarantees uniqueness of the key, and `extension.name` is historically used for specifying kind of extension, which is now implied by type, so that avoids confusion.

Reply:
I believe that nowadays, `extension.name` is intended to be used to specify the instance name, not the type of the extension.  This is reflected in the comment on that field:

https://github.com/envoyproxy/envoy/blob/c1bc5e78fa68b86236c8d6237e2db15ce1743459/api/envoy/config/core/v3/extension.proto#L22

This allows (for example) having multiple instances of the same type of extension under different names and then deciding which instance to use where.

Given that, I think using a map here would bloat the config unnecessarily by duplicating these instance names.

An alternative would be to change `ClusterSpecifierPlugin` to directly contain an `Any` field instead of using `TypedExtensionConfig`, so that we could put the name in the map here and not duplicate it.  But I think that we're trying to move toward using `TypedExtensionConfig` everywhere instead of bare `Any` fields, aren't we?

Reply:
Yes nowadays it is used to specify instance name.

I'm ok either way but we would need to do duplicate key check by using repeated.

Reply:
Yeah, I think doing the duplicate key check is fine.  The client implementation will need to put each key into a map anyway, so we can just check to make sure that the key is not already present right before we add it.

Reply:
Just a +1 to where we ended up (@markdroth asked for my opinion offline). I don't think it matters greatly, but DRY seems best (once again the weakness of protobuf type system strikes).